server/api.go

```diff
@@ -503,7 +552,7 @@ func (a *apiServer) subscribe(ctx context.Context, partition *partition,
 		headersBuf := make([]byte, 28)
 		for {
 			// TODO: this could be more efficient.
-			m, offset, timestamp, _, err := reader.ReadMessage(ctx, headersBuf)
+			m, offset, timestamp, _, err := reader.ReadMessage(ctx, headersBuf, partition.setReadonly)
```

Comment:
I do wonder if it would make more sense to mark the commitlog as `readonly` versus passing in a channel on read. I think I will explore that idea before we decide on this approach.

Reply:
It went with a channel because I felt it was the easiest way to interrupt the blocking read if a partition is set as read-only while there is a subscriber waiting for more data to arrive.

Another option could be to have a `commitLog`-internal channel that could be closed, instead of passing one? That would also solve the mutex-protected channel issue above.

Reply:
I have a branch where I mark the commitlog as readonly and interrupt readers waiting for data. Will push it up shortly.